File: tests/parity-check.py
"""Emit Python-canonical normalized forms for real verse text, for the JS parity
check. Run with the generator venv:
  cd ../ThaqalaynDataGenerator; .venv/Scripts/python.exe ../ThaqalaynSearch/tests/parity-check.py
Writes parity_py.json next to this file; parity-check.mjs then asserts the JS
normalizer produces identical output.
"""
import sys, os, json, glob

HERE = os.path.dirname(os.path.abspath(__file__))
GEN = os.path.abspath(os.path.join(HERE, "..", "..", "ThaqalaynDataGenerator"))
DATA = os.path.abspath(os.path.join(HERE, "..", "..", "ThaqalaynData"))
sys.path.insert(0, GEN)
sys.path.insert(0, os.path.join(GEN, "app"))

from arabic_normalization import normalize_arabic  # noqa: E402

out = {}
for f in sorted(glob.glob(os.path.join(DATA, "books", "al-amali-mufid", "**", "*.json"), recursive=True)):
    try:
        j = json.load(open(f, encoding="utf-8"))
    except Exception:
        continue
    if j.get("kind") != "verse_detail":
        continue
    v = j["data"]["verse"]
    out[v["path"]] = normalize_arabic(" ".join(v.get("text", [])))
    if len(out) >= 50:
        break

json.dump(out, open(os.path.join(HERE, "parity_py.json"), "w", encoding="utf-8"), ensure_ascii=False)
print("wrote", len(out), "normalized samples")

File: tests/parity-check.mjs
// Assert the JS Arabic normalizer matches the Python canonical on real verse
// text. Run parity-check.py first (it writes parity_py.json), then:
//   node tests/parity-check.mjs
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";
import { normalizeArabic } from "../lib/normalize-arabic.mjs";

const HERE = path.dirname(fileURLToPath(import.meta.url));
const DATA = path.resolve(HERE, "..", "..", "ThaqalaynData");
const py = JSON.parse(fs.readFileSync(path.join(HERE, "parity_py.json"), "utf8"));

let ok = 0, bad = 0;
const mismatches = [];
for (const [p, expected] of Object.entries(py)) {
  const file = path.join(DATA, "books", p.replace("/books/", "").replace(/:/g, "/") + ".json");
  const v = JSON.parse(fs.readFileSync(file, "utf8")).data.verse;
  const got = normalizeArabic((v.text || []).join(" "));
  if (got === expected) ok++;
  else { bad++; if (mismatches.length < 3) mismatches.push({ p, expected: expected.slice(0, 80), got: got.slice(0, 80) }); }
}
console.log(JSON.stringify({ ok, bad, mismatches }, null, 2));
process.exit(bad === 0 ? 0 : 1);
